### DISCUSSION TITLE
Can I put @NonBlocking everywhere on my resource endpoints using hibernate Panache?

### DISCUSSION THREAD

**Discussion by omar-zahid** [quarkusio/quarkus#25586] [Q&A] [answered]
Reading more from this blog: https://quarkus.io/blog/resteasy-reactive-smart-dispatch/
It seems like dispatching calls to I/O threads are more performant.
If I'm not using Mutiny, is it ok to put `@NonBlocking` annotations everywhere on all endpoints so I get 100 percent Quarkus REST Score Console.

``` java
   @PUT
   @Path("{id}")
   @Transactional
   @NonBlocking
   public Fruit update(Long id, Fruit fruit) {
        if (fruit== null || fruit.name== null) {
            throw new WebApplicationException("Fruit name was not set on request.", 422);
        }

        Fruit entity = Fruit.findById(id);

        if (entity == null) {
            throw new WebApplicationException("Fruit with id " + id + " was not found.", NOT_FOUND);
        }

        entity.name = fruit.name;
        return entity;

    }
```

[a comment by an automated account is not shown]

**Reply by yrodiere** [reply]
>  It seems like dispatching calls to I/O threads are more performant.

That's not what that blog post is telling you; or at the very least, this statement is taken out of context.

Reactive programming and dispatching calls to I/O threads may be more performant in some situations, but you need to write reactive code in the first place.

This blog post is telling you that Resteasy Reactive will work with both reactive code or imperative code, and is explaining you how to use Resteasy Reactive in each case. Most importantly, it's telling you that from Quarkus 2.2, Resteasy reactive should be able to detect whether your endpoint is blocking or non-blocking automatically, **without you needing to use `@Blocking` or `@NonBlocking`**.

> If I'm not using Mutiny, is it ok to put @NonBlocking annotations everywhere on all endpoints

`@NonBlocking` just "tells" Resteasy reactive that an endpoint is non-blocking, it doesn't magically change the underlying implementation. If you use `@NonBlocking` on your REST endpoint, then *you* are responsible for making sure your REST endpoint is non-blocking.

Apparently, you're using Hibernate ORM, not Hibernate Reactive. So, your endpoint is blocking. So no, it's not okay to use `@NonBlocking` on your (blocking) endpoint. At best putting `@NonBlocking` on your endpoint won't change anything, and more likely that will just degrade performance.

You could switch to Hibernate Reactive in order to make your endpoint non-blocking, but be aware that this is paradigm switch: you will have to write your application very differently (using reactive programming), and if you're not familiar with reactive programming this will be challenging. My personal opinion is you should only do this if you've identified a real need for your endpoint to be non-blocking.

I'll let the Resteasy Reactive experts expand on this and/or correct the mistakes in what I just wrote :) @geoand, @FroMage, @stuartwdouglas

@cescoffier may want to chime in, as he's the author of https://quarkus.io/blog/resteasy-reactive-smart-dispatch/

**Reply by omar-zahid** [reply]
Thanks @yrodiere This was helpful.

**Reply by geoand** [reply]
To build upon this excellent answer, the "overhead" that a blocking call introduces, is that the request needs to be offloaded from the IO thread to a worker thread. If all the JAX-RS method is doing is some super short in memory processing, this overhead is measurable.

Furthemore, `@NonBlocking` should **only** be used for operation that are super short - that is either operations that perform non-blocking IO or very short in memory processing.
Using a API that results in blocking IO calls - such as Hibernat ORM - should never be done in such methods, and Quarkus will indeed throw an exception for known errorneous usages (like the Hibernate ORM case).

**Comment by guitcastro** [comment]
Using @Nonblocking in a suspend function will be enough?
